daemon/state.go

```diff
@@ -353,7 +353,7 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) (resto
 			// The identity of a freshly restored endpoint is incomplete due to some
 			// parts of the identity not being marshaled to JSON. Hence we must set
 			// the identity even if has not changed.
-			ep.SetIdentity(identity)
+			ep.SetIdentity(identity, true)
```

Comment:
We might have already called `SetIdentity(oldIdentity, true`) in the `ParseEndpoint()` path in `pkg/endpoint`, then set a new, different identity here. In that case we should actually remove a reference to the old identity.

Reply:
Thanks, did not notice that. Planning to make `ParseEndpoint()` NOT call `SetIdentity()` at all, but ensure that `ep.SecurityIdentity` is non-nil:
```
diff --git a/pkg/endpoint/endpoint.go b/pkg/endpoint/endpoint.go
index 2dde1a4c89..c560343f28 100644
--- a/pkg/endpoint/endpoint.go
+++ b/pkg/endpoint/endpoint.go
@@ -1086,11 +1086,12 @@ func ParseEndpoint(owner regeneration.Owner, strEp string) (*Endpoint, error) {
 		ep.Status = NewEndpointStatus()
 	}
 
+	// Make sure the endpoint has an identity, using the 'init' identity if none.
 	if ep.SecurityIdentity == nil {
-		ep.SetIdentity(identityPkg.LookupReservedIdentity(identityPkg.ReservedIdentityInit), true)
-	} else {
-		ep.SecurityIdentity.Sanitize()
+		ep.SecurityIdentity = identityPkg.LookupReservedIdentity(identityPkg.ReservedIdentityInit)
 	}
+	ep.SecurityIdentity.Sanitize()
+
 	ep.UpdateLogger(nil)
 
 	ep.SetStateLocked(StateRestoring, "Endpoint restoring")
```

This way the state of the endpoint's identity is consistent after `ParseIdentity()`: not registered with the identitymanager yet.